extensions/shared/external-media/media-browser/index.js

```diff
@@ -99,36 +104,92 @@ function MediaBrowser( props ) {
 		);
 	};
 
+	/**
+	 * Counts how many items are in a row by checking how many of the grid's child items have a matching offsetTop.
+	 */
+	const checkColumns = () => {
+		let perRow = 1;
+
+		const items = gridEl.current.children;
+
+		if ( items.length > 0 ) {
+			const firstOffset = items[ 0 ].offsetTop;
+
+			// Check how many items have a matching offsetTop.
+			// This will give us the total number of items in a row.
+			while ( perRow < items.length && items[ perRow ].offsetTop === firstOffset ) {
+				++perRow;
+			}
+		}
+
+		setColumns( perRow );
+	};
+
+	useEffect( () => {
+		checkColumns();
+	}, [ media ] );
+
+	const handleArrowKeysNavigation = ( keyCode, index ) => {
+		switch ( keyCode ) {
+			case LEFT:
+				if ( index >= 1 ) {
+					setFocused( index - 1 );
+				}
+				break;
+			case RIGHT:
+				if ( index < media.length ) {
+					setFocused( index + 1 );
+				}
+				break;
+			case UP:
+				if ( index >= columns ) {
+					setFocused( index - columns );
+				}
+				break;
+			case DOWN:
+				if ( index < media.length - columns ) {
+					setFocused( index + columns );
+				}
+				break;
+		}
+	};
+
 	return (
 		<div className={ wrapper }>
-			<ul className={ classes }>
+			<ul ref={ gridEl } className={ classes }>
 				{ media.map( ( item, index ) => (
 					<MediaItem
 						item={ item }
 						key={ item.ID }
-						onClick={ onSelectImage }
-						focusOnMount={ !! prevMediaCount.current && index === prevMediaCount.current }
+						onClick={ image => {
+							onSelectImage( image );
+							setFocused( index );
+						} }
+						onKeyDown={ event => {
+							handleArrowKeysNavigation( event.keyCode, index );
```

Comment:
I don't think we should attach the handler to every item. This can cause performance issues, especially when paired with the "Load More" button which will create a new handler for every item that will be added. Instead, we should create a single handler on the parent and delegate the event.

Reply:
is it possible to use the useCallback hook?

Reply:
So my instinct is to say, yes delegation is better. Especially if we anticipate there are a tonne of media items. However, I'm not sure how much of an impact this will actually make.

Also there could be other knock-on issues caused by attempting this refactoring. For example, how will you get access to the `index` value if you attach the handler to the `<ul>` for example? 

Perhaps another option could be to use `useCallback` to memorize the result of the function for a given set of params. As per `useMemo` the function...

> ...will only recompute the memoized value when one of the dependencies has changed. 

...and therefore...

> This optimization helps to avoid expensive calculations on every render.

Reply:
> how will you get access to the index value if you attach the handler to the <ul> for example?

The index value is not necessary in this case, this works well based on the `focused` value only, as in https://github.com/Automattic/jetpack/pull/16118/commits/5fa86c8706f8e4719fb773999dbc8d03ae82221d#diff-f59fe14d5eeda67bd102566e399a16d4R104-R127

Reply:
Overall though I doubt the performance will be impacted much either way. Inline functions are very cheap.

Reply:
I guess if we agree that it **will** have a performance impact (no matter how big) we still should avoid it if possible. In this case, the solution is really just moving that handler to the parent so I think we still should do that.

Reply:
While I still don't know React very well, at this point I've read plenty of Gutenberg and Jetpack React code. I haven't seen an instance where someone adds the keydown listener or click listener to the parent. It appears as though React pools events and that they are [synthetic events](https://reactjs.org/docs/events.html), so I don't think we need to worry about the performance impact.

Also, I added the event listener there because there was already a click listener and keydown listener there, so this code isn't actually adding any new event listener (as far as I know).

Reply:
Cleaned up & consolidated handlers for clarity in https://github.com/Automattic/jetpack/pull/16118/commits/827298d31cc534ab62bcf7fb3541cec55d98460a.